File: packages/python/rails_ai_build/tools.py
"""Agent tools for codebase interaction."""

from __future__ import annotations

import json
import re
import subprocess
from abc import ABC, abstractmethod
from pathlib import Path
from typing import Any

from rails_ai_build.config import get_config

BLOCKED_SHELL = [
    r"\brm\s+-rf\s+/\b",
    r"\bmkfs\b",
    r"\bdd\s+if=",
]


class BaseTool(ABC):
    name: str = ""
    description: str = ""
    parameters: dict = {}

    def __init__(self, workspace: Path):
        self.workspace = workspace.resolve()

    def definition(self) -> dict:
        return {
            "name": self.name,
            "description": self.description,
            "parameters": self.parameters,
        }

    def resolve_path(self, path: str) -> Path:
        full = (self.workspace / path.lstrip("/")).resolve()
        if not str(full).startswith(str(self.workspace)):
            raise PermissionError(f"Path escapes workspace: {path}")
        return full

    @abstractmethod
    def execute(self, args: dict[str, Any]) -> dict[str, Any]:
        ...


class ReadFileTool(BaseTool):
    name = "read_file"
    description = "Read the contents of a file in the workspace."
    parameters = {
        "type": "object",
        "properties": {
            "path": {"type": "string"},
            "offset": {"type": "integer"},
            "limit": {"type": "integer"},
        },
        "required": ["path"],
    }

    def execute(self, args):
        path = self.resolve_path(args["path"])
        if not path.is_file():
            return {"error": f"File not found: {args['path']}"}
        lines = path.read_text().splitlines()
        offset = max((args.get("offset") or 1) - 1, 0)
        limit = args.get("limit")
        selected = lines[offset : offset + limit] if limit else lines[offset:]
        numbered = [f"{offset + i + 1}|{line}" for i, line in enumerate(selected)]
        return {"path": args["path"], "content": "\n".join(numbered), "total_lines": len(lines)}


class WriteFileTool(BaseTool):
    name = "write_file"
    description = "Create or overwrite a file in the workspace."
    parameters = {
        "type": "object",
        "properties": {
            "path": {"type": "string"},
            "content": {"type": "string"},
        },
        "required": ["path", "content"],
    }

    def execute(self, args):
        path = self.resolve_path(args["path"])
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(args["content"])
        return {"path": args["path"], "bytes_written": len(args["content"]), "status": "written"}


class GrepTool(BaseTool):
    name = "grep"
    description = "Search for a pattern in workspace files."
    parameters = {
        "type": "object",
        "properties": {
            "pattern": {"type": "string"},
            "path": {"type": "string"},
            "glob": {"type": "string"},
            "case_insensitive": {"type": "boolean"},
        },
        "required": ["pattern"],
    }

    def execute(self, args):
        pattern = args["pattern"]
        flags = re.IGNORECASE if args.get("case_insensitive") else 0
        regex = re.compile(pattern, flags)
        base = self.resolve_path(args["path"]) if args.get("path") else self.workspace
        glob = args.get("glob", "**/*")
        matches = []

        for file in base.glob(glob):
            if not file.is_file() or "/.git/" in str(file):
                continue
            try:
                for i, line in enumerate(file.read_text().splitlines(), 1):
                    if regex.search(line):
                        rel = file.relative_to(self.workspace)
                        matches.append(f"{rel}:{i}:{line}")
                        if len(matches) >= 100:
                            break
            except (UnicodeDecodeError, PermissionError):
                continue
            if len(matches) >= 100:
                break

        return {"pattern": pattern, "matches": matches, "count": len(matches)}


class ListFilesTool(BaseTool):
    name = "list_files"
    description = "List files and directories in the workspace."
    parameters = {
        "type": "object",
        "properties": {
            "path": {"type": "string"},
            "glob": {"type": "string"},
            "max_results": {"type": "integer"},
        },
    }

    def execute(self, args):
        base = self.resolve_path(args["path"]) if args.get("path") else self.workspace
        glob = args.get("glob", "**/*")
        max_results = args.get("max_results", 200)
        entries = sorted(
            str(p.relative_to(self.workspace))
            for p in base.glob(glob)
            if p.is_file() and "/.git/" not in str(p)
        )[:max_results]
        return {"path": args.get("path", "."), "entries": entries, "count": len(entries)}


class ShellTool(BaseTool):
    name = "shell"
    description = "Execute a shell command in the workspace."
    parameters = {
        "type": "object",
        "properties": {
            "command": {"type": "string"},
            "timeout": {"type": "integer"},
        },
        "required": ["command"],
    }

    def execute(self, args):
        command = args["command"].strip()
        for pattern in BLOCKED_SHELL:
            if re.search(pattern, command):
                return {"error": "Command blocked for safety"}
        timeout = args.get("timeout") or get_config().shell_timeout
        try:
            result = subprocess.run(
                command,
                shell=True,
                cwd=str(self.workspace),
                capture_output=True,
                text=True,
                timeout=timeout,
            )
            return {
                "command": command,
                "exit_code": result.returncode,
                "stdout": result.stdout[:50_000],
                "stderr": result.stderr[:10_000],
            }
        except subprocess.TimeoutExpired:
            return {"error": f"Command timed out after {timeout}s", "command": command}


TOOLS: dict[str, type[BaseTool]] = {
    "read_file": ReadFileTool,
    "write_file": WriteFileTool,
    "grep": GrepTool,
    "list_files": ListFilesTool,
    "shell": ShellTool,
}


def build_tools(workspace: Path) -> dict[str, BaseTool]:
    allowed = get_config().allowed_tools
    return {name: TOOLS[name](workspace) for name in allowed if name in TOOLS}


def tool_definitions() -> list[dict]:
    workspace = get_config().workspace_root
    return [build_tools(workspace)[name].definition() for name in get_config().allowed_tools if name in TOOLS]


def execute_tool(name: str, arguments: dict, workspace: Path) -> dict:
    if name not in get_config().allowed_tools:
        return {"error": f"Tool not allowed: {name}"}
    tool = TOOLS[name](workspace)
    try:
        return tool.execute(arguments)
    except Exception as e:
        return {"error": str(e)}

File: packages/javascript/src/tools.ts
import { exec } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { promisify } from "node:util";
import { getConfig } from "./config.js";
import type { ToolDefinition } from "./providers.js";

const execAsync = promisify(exec);

const BLOCKED_SHELL = [/\brm\s+-rf\s+\//, /\bmkfs\b/, /\bdd\s+if=/];

export interface Tool {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
  execute(args: Record<string, unknown>): Promise<Record<string, unknown>>;
}

function resolvePath(workspace: string, filePath: string): string {
  const full = path.resolve(workspace, filePath.replace(/^\//, ""));
  if (!full.startsWith(path.resolve(workspace))) {
    throw new Error(`Path escapes workspace: ${filePath}`);
  }
  return full;
}

export function createTools(workspace: string): Record<string, Tool> {
  return {
    read_file: {
      name: "read_file",
      description: "Read the contents of a file in the workspace.",
      parameters: {
        type: "object",
        properties: {
          path: { type: "string" },
          offset: { type: "integer" },
          limit: { type: "integer" },
        },
        required: ["path"],
      },
      async execute(args) {
        const full = resolvePath(workspace, args.path as string);
        if (!fs.existsSync(full)) return { error: `File not found: ${args.path}` };
        const lines = fs.readFileSync(full, "utf-8").split("\n");
        const offset = Math.max(((args.offset as number) || 1) - 1, 0);
        const limit = args.limit as number | undefined;
        const selected = limit ? lines.slice(offset, offset + limit) : lines.slice(offset);
        const numbered = selected.map((line, i) => `${offset + i + 1}|${line}`);
        return { path: args.path, content: numbered.join("\n"), total_lines: lines.length };
      },
    },

    write_file: {
      name: "write_file",
      description: "Create or overwrite a file in the workspace.",
      parameters: {
        type: "object",
        properties: {
          path: { type: "string" },
          content: { type: "string" },
        },
        required: ["path", "content"],
      },
      async execute(args) {
        const full = resolvePath(workspace, args.path as string);
        fs.mkdirSync(path.dirname(full), { recursive: true });
        fs.writeFileSync(full, args.content as string);
        return { path: args.path, bytes_written: (args.content as string).length, status: "written" };
      },
    },

    grep: {
      name: "grep",
      description: "Search for a pattern in workspace files.",
      parameters: {
        type: "object",
        properties: {
          pattern: { type: "string" },
          path: { type: "string" },
          case_insensitive: { type: "boolean" },
        },
        required: ["pattern"],
      },
      async execute(args) {
        const flags = args.case_insensitive ? "i" : "";
        const regex = new RegExp(args.pattern as string, flags);
        const base = args.path ? resolvePath(workspace, args.path as string) : workspace;
        const matches: string[] = [];

        function walk(dir: string) {
          if (matches.length >= 100) return;
          for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
            if (entry.name === ".git" || entry.name === "node_modules") continue;
            const full = path.join(dir, entry.name);
            if (entry.isDirectory()) walk(full);
            else if (entry.isFile()) {
              try {
                const lines = fs.readFileSync(full, "utf-8").split("\n");
                lines.forEach((line, i) => {
                  if (regex.test(line)) {
                    matches.push(`${path.relative(workspace, full)}:${i + 1}:${line}`);
                  }
                });
              } catch { /* skip binary/unreadable */ }
            }
            if (matches.length >= 100) break;
          }
        }

        walk(base);
        return { pattern: args.pattern, matches, count: matches.length };
      },
    },

    list_files: {
      name: "list_files",
      description: "List files in the workspace.",
      parameters: {
        type: "object",
        properties: {
          path: { type: "string" },
          max_results: { type: "integer" },
        },
      },
      async execute(args) {
        const base = args.path ? resolvePath(workspace, args.path as string) : workspace;
        const max = (args.max_results as number) || 200;
        const entries: string[] = [];

        function walk(dir: string) {
          if (entries.length >= max) return;
          for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
            if (entry.name === ".git" || entry.name === "node_modules") continue;
            const full = path.join(dir, entry.name);
            if (entry.isDirectory()) walk(full);
            else entries.push(path.relative(workspace, full));
            if (entries.length >= max) break;
          }
        }

        walk(base);
        return { path: args.path || ".", entries: entries.sort(), count: entries.length };
      },
    },

    shell: {
      name: "shell",
      description: "Execute a shell command in the workspace.",
      parameters: {
        type: "object",
        properties: {
          command: { type: "string" },
          timeout: { type: "integer" },
        },
        required: ["command"],
      },
      async execute(args) {
        const command = (args.command as string).trim();
        if (BLOCKED_SHELL.some((p) => p.test(command))) {
          return { error: "Command blocked for safety" };
        }
        const timeout = (args.timeout as number) || getConfig().shellTimeout;
        try {
          const { stdout, stderr } = await execAsync(command, {
            cwd: workspace,
            timeout: timeout * 1000,
            maxBuffer: 50_000,
          });
          return { command, exit_code: 0, stdout, stderr };
        } catch (err: unknown) {
          const e = err as { code?: number; stdout?: string; stderr?: string; killed?: boolean };
          if (e.killed) return { error: `Command timed out after ${timeout}s`, command };
          return { command, exit_code: e.code ?? 1, stdout: e.stdout || "", stderr: e.stderr || "" };
        }
      },
    },
  };
}

export function toolDefinitions(workspace: string): ToolDefinition[] {
  const all = createTools(workspace);
  return getConfig().allowedTools.filter((n) => all[n]).map((n) => ({
    name: all[n].name,
    description: all[n].description,
    parameters: all[n].parameters,
  }));
}

export async function executeTool(
  name: string,
  args: Record<string, unknown>,
  workspace: string
): Promise<Record<string, unknown>> {
  const all = createTools(workspace);
  if (!getConfig().allowedTools.includes(name)) return { error: `Tool not allowed: ${name}` };
  const tool = all[name];
  if (!tool) return { error: `Unknown tool: ${name}` };
  try {
    return await tool.execute(args);
  } catch (err: unknown) {
    return { error: err instanceof Error ? err.message : String(err) };
  }
}
